ding/policy/ibc.py

```diff
@@ -0,0 +1,187 @@
+from typing import Dict, Any, List, Tuple
+from collections import namedtuple
+from easydict import EasyDict
+
+import torch
+import torch.nn.functional as F
+
+from ding.model import model_wrap
+from ding.torch_utils import to_device
+from ding.utils.data import default_collate, default_decollate
+from ding.utils import POLICY_REGISTRY
+from .bc import BehaviourCloningPolicy
+from ding.model.template.ebm import create_stochastic_optimizer
+from ding.model.template.ebm import StochasticOptimizer, MCMC, AutoRegressiveDFO
+from ding.torch_utils import unsqueeze_repeat
+from ding.utils import EasyTimer
+
+
+@POLICY_REGISTRY.register('ibc')
+class IBCPolicy(BehaviourCloningPolicy):
+    r"""
+    Overview:
+        Implicit Behavior Cloning
+        https://arxiv.org/abs/2109.00137.pdf
+
+    Config:
+        TODO
+    """
+
+    config = dict(
+        type='ibc',
+        cuda=False,
+        on_policy=False,
+        continuous=True,
+        learn=dict(
+            train_epoch=30,
+            batch_size=256,
+            multi_gpu=False,
+            optim=dict(
+                learning_rate=1e-5,
+                weight_decay=0.0,
+                beta1=0.9,
+                beta2=0.999,
+            ),
+        ),
+        collect=dict(normalize_states=True, ),
+        eval=dict(evaluator=dict(eval_freq=10000, )),
+    )
+
+    def default_model(self) -> Tuple[str, List[str]]:
+        return 'ebm', ['ding.model.template.ebm']
+
+    def _init_learn(self):
+        self._timer = EasyTimer(cuda=self._cfg.cuda)
+        self._sync_timer = EasyTimer(cuda=self._cfg.cuda)
+        optim_cfg = self._cfg.learn.optim
+        self._optimizer = torch.optim.Adam(
+            self._model.parameters(),
+            lr=optim_cfg.learning_rate,
+            weight_decay=optim_cfg.weight_decay,
+            betas=(optim_cfg.beta1, optim_cfg.beta2),
+        )
+        self._stochastic_optimizer: StochasticOptimizer = \
+            create_stochastic_optimizer(self._cfg.model.stochastic_optim)
```

Comment:
we can set device for stochastic_optimizer here, rather than use config

Reply:
changed